### PR TITLE
Use None on write_autoconf for default header and add it as parameter on genconfig

### PR DESCRIPTION
**Pull request by leandrolanzieri**
This PR modifies the `write_autoconf` function so that the default value for the `header` parameter is None, if that is the case then the default header title is added to the header file.

This allows things like the proposed modification on the `genconfig` script, where `write_autoconf` can be called with None when no specific header file title is needed. The behaviour of the function does not change when `header` is not specified.

### CODE REVIEW

**Comment by ulfalizer** [comment]
I wonder if this should be customizable via environment variables instead. Having to add separate options to each tool that writes configuration files and headers seems like a pain.

Could have e.g. `KCONFIG_CONFIG_HEADER` and `KCONFIG_HEADER_HEADER`. Thoughts?

**Comment by cgundogan** [comment]
>Could have e.g. KCONFIG_CONFIG_HEADER and KCONFIG_HEADER_HEADER, that would kick in if header=None (with an ultimate fallback on the default string). Thoughts?

:+1: that would reduce command line clutter, escpecially for very long headers. However, being able to set the header explicitly via a command line option is also useful for setups, where modifying the environment is not that simple (testing, CI, ..).

What about supporting both options? With a preference of the command line option, if both are set? The environment variable would then represent the *default header*.

File: genconfig.py [hunk 1]
```diff
@@ -44,6 +44,12 @@ def main():
         help="Path for the generated header file (default: {})"
              .format(DEFAULT_HEADER_PATH))
 
+    parser.add_argument(
+        "--header-title",
+        metavar="HEADER_TITLE",
+        default=None,
+        help="Title added on the first line of the generated header file")
```
[the remark this answers is not in the document]
Comment (leandrolanzieri) [reply genconfig.py]: Hmm yes, title does not seem to be precise here. What about `--header-preamble`?

File: kconfiglib.py [hunk 2]
```diff
@@ -1370,8 +1369,11 @@ def write_autoconf(self, filename,
         header (default: "/* Generated by Kconfiglib (https://github.com/ulfalizer/Kconfiglib) */\n"):
```
[the remark this answers is not in the document]
Comment (leandrolanzieri) [reply kconfiglib.py]: I will change it

File: genconfig.py [hunk 3]
```diff
@@ -44,6 +44,12 @@ def main():
         help="Path for the generated header file (default: {})"
              .format(DEFAULT_HEADER_PATH))
 
+    parser.add_argument(
+        "--header-title",
+        metavar="HEADER_TITLE",
+        default=None,
```
[the remark this answers is not in the document]
Comment (leandrolanzieri) [reply genconfig.py]: Removed

File: genconfig.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cgundogan) [reply genconfig.py]: `--header-preamble` looks like a good choice for boilerplate that is added in front of the actual header file  :+1: Alternatively `--header-prologue`, although I would prefer the former option.

**Comment by ulfalizer** [comment]
Thanks for the changes!

> > Could have e.g. KCONFIG_CONFIG_HEADER and KCONFIG_HEADER_HEADER, that would kick in if header=None (with an ultimate fallback on the default string). Thoughts?
> 
>  that would reduce command line clutter, escpecially for very long headers. However, being able to set the header explicitly via a command line option is also useful for setups, where modifying the environment is not that simple (testing, CI, ..).
> 
> What about supporting both options? With a preference of the command line option, if both are set? The environment variable would then represent the _default header_.

One issue with the command-line options is that I'd want all commands to support them, for consistency. Would need an `allyesconfig.py --config-header="# foo bar"` for example. That'd make all tools grow a bit, unless it could be handled with common code somehow.

I'll implement just the environment variable thing for now. I could maybe add some command-line stuff later on as well if it turns out to be messy though.

File: genconfig.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply genconfig.py]: Yeah, preamble sounds pretty good. Thanks for the suggestion!

**Comment by ulfalizer** [comment]
I'll add this this weekend at the latest. Busy times with some devicetree stuff atm. :)

Gonna add `KCONFIG_AUTOHEADER` as another fallback for the header path at the same time. It's the name used by the C tools. The env. var. for the header preamble should probably be called `KCONFIG_AUTOHEADER_PREAMBLE`, for consistency.

**Comment by leandrolanzieri** [comment]
@ulfalizer I implemented the proposed solution. I assumed the default autoheader path to be the one used in `genconfig`.

**Comment by leandrolanzieri** [comment]
> * If `--header-path` is not passed, but `KCONFIG_AUTOHEADER` is set, then `None` is passed to `write_autoconf()` to use the path from the environment

The only downside that I see on this is that now the environmental variable has to be checked in `genconfig.py` as well.

**Comment by ulfalizer** [comment]
Added now, in https://github.com/ulfalizer/Kconfiglib/commit/de45874719772a40f1d8d244e2f5a6c6036415ac and https://github.com/ulfalizer/Kconfiglib/commit/bb3be6ee9793f23c275d318f58cb6cf21391b371. Will wait for feedback before pushing out a release.

I removed the Kconfiglib advertising, because it didn't feel worthwhile to add extra code and variables just to keep it. The default is now no header. :)

It's a bit gnarly that you have to add the trailing newline yourself in the common case of just adding a short comment at the top, though it has some nice side effects, like the header disappearing completely with `KCONFIG_AUTOHEADER_HEADER="$FOO"` when `$FOO` isn't set.

**Comment by ulfalizer** [comment]
Went with "header" instead of "preamble" for consistency with the parameter names. Hopefully it won't be confused with header as in header file.

**Comment by leandrolanzieri** [comment]
> I removed the Kconfiglib advertising, because it didn't feel worthwhile to add extra code and variables just to keep it. The default is now no header. :)

Sounds good.
@ulfalizer thanks for addressing this. Closing now

**Pull request closed without merging**